### DISCUSSION TITLE
xmemdump completion status

### DISCUSSION THREAD

**Discussion by eddietorial** [CrowdStrike/psfalcon#297] [Q&A]
How do I check the status of xmemdump. Is there a way to get a confirmation that it is in progress or has completed?

The output is sent to a network share as local drive output is not an option in our environment. However, xmemdump takes over an hour to complete, so knowing that it has successfully completed or has failed would be useful.

**Comment by bk-cs** [comment]
There's no progress output when creating the `xmemdump` output, so my guess is that you'll need a script that checks for the presence of the output file, and determines whether or not it's complete. Maybe run something like `Invoke-FalconRtr` to start the memdump generation, record the path of the dump file you create, check for the file to verify that it's no longer locked (using RTR), and once it's available, `get` it?

I'll try and put together an example in my spare time, but the next couple of weeks are going to be pretty busy. If you figure it out beforehand, please reply with your solution!

**Reply by eddietorial** [reply]
This worked for me to chek if the file is locked...

https://stackoverflow.com/questions/24992681/powershell-check-if-a-file-is-locked
```
function Test-FileLock {
  param (
    [parameter(Mandatory=$true)][string]$Path
  )

  $oFile = New-Object System.IO.FileInfo $Path

  if ((Test-Path -Path $Path) -eq $false) {
    return $false
  }

  try {
    $oStream = $oFile.Open([System.IO.FileMode]::Open, [System.IO.FileAccess]::ReadWrite, [System.IO.FileShare]::None)

    if ($oStream) {
      $oStream.Close()
    }
    $false
  } catch {
    # file is locked by a process.
    return $true
  }
}
```

One problem solved. The next is a token request failure within 30 minutes of the script executing xmemdump.

```
Authorization token request failed.
At C:\Program Files\WindowsPowerShell\Modules\PSFalcon\2.2.4\Public\oauth2.ps1:186 char:17
+                 throw 'Authorization token request failed.'
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Authorization token request failed.:String) [], RuntimeExce 
   ption
    + FullyQualifiedErrorId : Authorization token request failed.

```

I was under the impression that as long as the PowerShell session (or script) stays open, we would not have to worry about an expiring token.
https://www.reddit.com/r/crowdstrike/comments/qdmoat/psfalcon_powershell_token_expiration/

`$run_xmemmdump = Invoke-FalconRtr -Command xmemdump -Arguments "Complete $memory_dump_file" -HostId $host_id`

In an attempt to keep the token alive, I have a loop where while the file lock is TRUE, run `Invoke-FalconRtr -Command ls -Arguments "$memory_dump_file" -HostId $host_id` 

I have also tried FalconSession...
```
$start_session = Start-FalconSession -Id $host_id
$session_id = $start_session.session_id

$run_xmemmdump = Invoke-FalconAdminCommand -Command xmemdump -Argument "Complete $memory_dump_file" -SessionId $session_id
```
 and then loop `$update_session = Update-FalconSession -HostId $host_id` every minute

But these do not work as within 30 minutes the script terminates with the `Authorization token request failed`.

On a positive note, the xmemdump continues to run successfully as I can see the output on the network share. 

However, as the script terminates prematurely, it does not help to know for sure that the xmemdump was unsuccessful or do any further tasks on the script.

**Reply by eddietorial** [reply]
That's odd. Not sure if there is anything that's affecting the connection.

I've tested this at least 20 times and in my observation - after the `xmemdump` is initiated and the token throws an error within 30 minutes,  the memory dump continues to send data to the network share. So this connection is definitely not killed.

Also, for debugging, I have periodic stdout from `ls` running on loop, assuming that this would keep the token alive. I haven't seen errors with the `ls` output. Just the abrupt token error which breaks the script.

In my environment, `xmemdump` takes over 30 minutes for RAM over 8GB even if it is output locally to `C:\`. We do not have a pre-mapped network share dedicated for memory dumps on all workstations. So the mapping and unmapping of a network drive on the suspect machine before and after running `xmemdump` is essential. 

I'll just  have to take another approach. Thank you for your time. Much appreciated!

**Reply by bk-cs** [reply]
Did you try with my script? It will check every 30 seconds to determine whether the file is done being created.

PSFalcon is designed to get a token if there isn't a valid one available whenever it makes a request. If the token is due to expire within 1 minute (or has already expired), it makes the token request. If there's an error generated by the API, you'll get that error message back.

The only time when `Authorization token request failed` is returned is when the token request is made, but nothing comes back from the API. I've never seen this happen unless the PowerShell session is unable to make a connection to the API... and I've definitely never seen it happen after an initial token request was successful.

* Are you running this script on a physical machine in a regular PowerShell session?
* Is it connected directly to the internet? 
* Could anything be disconnecting you from the internet, or interrupting that connection over time?

**Reply by eddietorial** [reply]
Are you running this script on a physical machine in a regular PowerShell session?
-- Yes, physical machine, running Windows 10 Enterprise and regular PowerShell session (PSVersion 5.1.19041.2673)

Is it connected directly to the internet?
-- It's on our internal network and has outbound internet access. There are network security layers (firewall, proxy) so it isn't direct.

Could anything be disconnecting you from the internet, or interrupting that connection over time?
-- I'm seeing these in our winevent logs _just before_ the script fails with `Authorization token request failed`.  

`Message=Error Message = Invalid JSON primitive: True` for `ConvertFrom-Json` from the script `C:\Program Files\WindowsPowerShell\Modules\PSFalcon\2.2.4\Public\psf-real-time-response.ps1` 

and from our proxy logs ` HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)` 

Here's a snippet from our SIEM logs, and from your script.

SIEM (winevent logs)

```
04/14/2023 03:14:09 AM
LogName=Microsoft-Windows-PowerShell/Operational
EventCode=4100
EventType=3
[SNIP]
SourceName=Microsoft-Windows-PowerShell
Type=Warning
RecordNumber=24164
Keywords=None
TaskCategory=Executing Pipeline
OpCode=To be used when an exception is raised
Message=Error Message = Invalid JSON primitive: True.
Fully Qualified Error ID = System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Context:
        Severity = Warning
[SNIP]
        Command Name = ConvertFrom-Json
        Command Type = Cmdlet
        Script Name = C:\Program Files\WindowsPowerShell\Modules\PSFalcon\2.2.4\Public\psf-real-time-response.ps1
[SNIP]


2023-04-14 03:15:12	
LogName=Microsoft-Windows-PowerShell/Operational
EventCode=4100
EventType=3
[SNIP]
SourceName=Microsoft-Windows-PowerShell
Type=Warning
RecordNumber=24167
Keywords=None
TaskCategory=Executing Pipeline
OpCode=To be used when an exception is raised
Message=Error Message = Authorization token request failed.
Fully Qualified Error ID = Authorization token request failed.

Context:
        Severity = Warning
[SNIP]
        Script Name = C:\Program Files\WindowsPowerShell\Modules\PSFalcon\2.2.4\Public\oauth2.ps1
[SNIP]
```


SIEM (proxy logs)
```
_time	activity	useragent	uri_path	fullurl	x_client_ssl_err
2023-04-14 00:12:12	 	 	crowdstrike-psfalcon/2.2.4	/real-time-response/combined/batch-init-session/v1?timeout=30&host_timeout_duration=25s	https://api.us-2.crowdstrike.com/real-time-response/combined/batch-init-session/v1?timeout=30&host_timeout_duration=25s?timeout=30&host_timeout_duration=25s	 
2023-04-14 00:12:13	 	 	crowdstrike-psfalcon/2.2.4	/real-time-response/combined/batch-admin-command/v1?timeout=30&host_timeout_duration=25s	https://api.us-2.crowdstrike.com/real-time-response/combined/batch-admin-command/v1?timeout=30&host_timeout_duration=25s?timeout=30&host_timeout_duration=25s	 
2023-04-14 00:12:46	Login Attempt	crowdstrike-psfalcon/2.2.4
2023-04-14 00:12:46	 	 	crowdstrike-psfalcon/2.2.4	/oauth2/token	 	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)
	 	 	 	 

2023-04-14 03:14:39	 	 	crowdstrike-psfalcon/2.2.4	/real-time-response/combined/batch-init-session/v1?timeout=30&host_timeout_duration=25s	https://api.us-2.crowdstrike.com/real-time-response/combined/batch-init-session/v1?timeout=30&host_timeout_duration=25s?timeout=30&host_timeout_duration=25s	 
2023-04-14 03:14:40	 	 	crowdstrike-psfalcon/2.2.4	/real-time-response/combined/batch-admin-command/v1?timeout=30&host_timeout_duration=25s	https://api.us-2.crowdstrike.com/real-time-response/combined/batch-admin-command/v1?timeout=30&host_timeout_duration=25s?timeout=30&host_timeout_duration=25s	 
2023-04-14 03:15:12	Login Attempt	crowdstrike-psfalcon/2.2.4	 
2023-04-14 03:15:12	 	 	crowdstrike-psfalcon/2.2.4	/oauth2/token	 	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)
```

Your script stdout
```
VERBOSE: [ApiClient.Invoke] 201: Created
VERBOSE: [ApiClient.Invoke] Connection=keep-alive, Strict-Transport-Security=max-age=15724800; includeSubDomai
ns, max-age=31536000; includeSubDomains, X-Cs-Region=us-2, X-Cs-Traceid=f531d389-978d-4688-a780-889e23aabba5, 
X-Ratelimit-Limit=6000, X-Ratelimit-Remaining=5996, Date=Fri, 14 Apr 2023 08:14:09 GMT, Server=nginx
VERBOSE: [ApiClient.Invoke] POST https://api.us-2.crowdstrike.com/real-time-response/combined/batch-init-sessi
on/v1?timeout=30&host_timeout_duration=25s
VERBOSE: [ApiClient.Invoke] ContentType=application/json, Accept=application/json
VERBOSE: [ApiClient.Invoke] {"host_ids":["c6bd9d4e2f924e27b0833c6fb23d1030"]}
VERBOSE: [ApiClient.Invoke] 201: Created
VERBOSE: [ApiClient.Invoke] Connection=keep-alive, Strict-Transport-Security=max-age=15724800; includeSubDomai
ns, max-age=31536000; includeSubDomains, X-Cs-Region=us-2, X-Cs-Traceid=c2902f3d-16b3-4519-87d4-04227ee6ae65, 
X-Ratelimit-Limit=6000, X-Ratelimit-Remaining=5996, Date=Fri, 14 Apr 2023 08:14:40 GMT, Server=nginx
VERBOSE: [ApiClient.Invoke] POST https://api.us-2.crowdstrike.com/real-time-response/combined/batch-admin-comm
and/v1?timeout=30&host_timeout_duration=25s
VERBOSE: [ApiClient.Invoke] ContentType=application/json, Accept=application/json
VERBOSE: [ApiClient.Invoke] {"command_string":"runscript -Raw=```$File = New-Object System.IO.FileInfo X:\\202
3-04-14-test-mem.dmp\r\ntry {\r\n    $Stream = $File.Open([System.IO.FileMode]::Open,[System.IO.FileAccess]::R
eadWrite,[System.IO.FileShare]::None)\r\n    if ($Stream) { $Stream.Close() }\r\n    $false\r\n} catch {\r\n  
  $true\r\n}```","batch_id":"fc9e8bac-93bb-4ee1-9c63-5b06d0c36035","base_command":"runscript"}
VERBOSE: [ApiClient.Invoke] 201: Created
VERBOSE: [ApiClient.Invoke] Connection=keep-alive, Strict-Transport-Security=max-age=15724800; includeSubDomai
ns, max-age=31536000; includeSubDomains, X-Cs-Region=us-2, X-Cs-Traceid=2f4a47b2-70e1-4e29-8b72-4488d5b80627, 
X-Ratelimit-Limit=6000, X-Ratelimit-Remaining=5995, Date=Fri, 14 Apr 2023 08:14:42 GMT, Server=nginx
VERBOSE: [ApiClient.Invoke] POST https://api.us-2.crowdstrike.com/oauth2/token
VERBOSE: [ApiClient.Invoke] ContentType=application/x-www-form-urlencoded, Accept=application/json
Authorization token request failed.
At C:\Program Files\WindowsPowerShell\Modules\PSFalcon\2.2.4\Public\oauth2.ps1:186 char:17
+                 throw 'Authorization token request failed.'
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Authorization token request failed.:String) [], RuntimeExce 
   ption
    + FullyQualifiedErrorId : Authorization token request failed.
```

**Reply by bk-cs** [reply]
The error message indicates that the response from the token API is not Json. I've seen this happen in rare occasions when an API is broken, but not predictably like you're seeing.

Is it possible that the proxy is intercepting and breaking the SSL connection used during the token refresh? I would assume the initial request would have problems too...

**Reply by eddietorial** [reply]
The initial request has no problems. The request after about 30 minutes always fails. Not sure how the proxy would be interfering with it as every other task with psfalcon runs without any trouble.


```
_time	                http_user_agent	                uri_path	x_client_ssl_err
2023-04-06 07:15:03	crowdstrike-psfalcon/2.2.4	/oauth2/token	 
2023-04-06 07:44:02	crowdstrike-psfalcon/2.2.4	/oauth2/token	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)

2023-04-10 07:34:36	crowdstrike-psfalcon/2.2.4	/oauth2/token	 
2023-04-10 08:03:36	crowdstrike-psfalcon/2.2.4	/oauth2/token	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)

2023-04-10 23:09:52	crowdstrike-psfalcon/2.2.4	/oauth2/token	 
2023-04-10 23:38:59	crowdstrike-psfalcon/2.2.4	/oauth2/token	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)

2023-04-11 06:53:06	crowdstrike-psfalcon/2.2.4	/oauth2/token	 
2023-04-11 07:22:06	crowdstrike-psfalcon/2.2.4	/oauth2/token	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)

2023-04-12 05:33:25	crowdstrike-psfalcon/2.2.4	/oauth2/token	 
2023-04-12 06:02:25	crowdstrike-psfalcon/2.2.4	/oauth2/token	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)

2023-04-12 06:47:57	crowdstrike-psfalcon/2.2.4	/oauth2/token	 
2023-04-12 07:16:57	crowdstrike-psfalcon/2.2.4	/oauth2/token	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)

2023-04-13 23:43:36	crowdstrike-psfalcon/2.2.4	/oauth2/token	 
2023-04-14 00:12:46	crowdstrike-psfalcon/2.2.4	/oauth2/token	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)

2023-04-14 02:45:55	crowdstrike-psfalcon/2.2.4	/oauth2/token	 
2023-04-14 03:15:12	crowdstrike-psfalcon/2.2.4	/oauth2/token	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)

2023-04-14 04:57:04	crowdstrike-psfalcon/2.2.4	/oauth2/token	 
2023-04-14 05:26:30	crowdstrike-psfalcon/2.2.4	/oauth2/token	HsFailure (error:1408F10B:SSL routines:ssl3_get_record:wrong version number)
```

**Reply by bk-cs** [reply]
Can you run it without the proxy?

**Reply by eddietorial** [reply]
Not an option in my environment. I have tested and know it does work without issues in a direct internet connection environment. But we're not going to be able to give it a direct connection for our internal use-case. The question being asked is, what is the script doing differently around the 30th minute before it fails? All psfalcon requests up to that point in time go through successfully without any blocks.

**Reply by eddietorial** [reply]
As the auto renew fails for me, I have my script manually request a new token.

However, if the current token is still active, issuing a `Request-FalconToken` gives me the same error as seen in our proxy logs.

So, I have to force a `Revoke-FalconToken`, before issuing a `Request-FalconToken` and that seems to work in my environment.

**Reply by bk-cs** [reply]
It seems to me that your proxy is doing something with the SSL connection, or the API is returning something that your proxy does not expect. If it doesn't happen when the proxy is removed, the module is not the problem.

When it automatically requests a new token at 30 minutes, it's running the exact same command as it is the first time -- `Request-FalconToken`.

**Reply by eddietorial** [reply]
So, when it automatically requests a new token at 30 minutes with `Request-FalconToken`, does it first test if the existing token exists? And if so, does it revoke it before the new token request?

I agree that there may be something happening on the wire that we're not seeing in the logs. However, I am able to reproduce the same error in the proxy logs without having to wait 30 minutes, just by running `Request-FalconToken` without revoking the existing token.

I revoke the existing token in my script before requesting a new token regardless of whether `((Test-FalconToken).Token -eq $true)` else the new token request fails just at it would in the predefined auto-check at 30 minutes.

In my environment, `xmemdump` is the only RTR process that takes over 30 minutes to output to a network share, I run the new token request every 10 minutes while checking the file lock. Sure, this delays the confirmation of the `xmemdump` output completion but at least the script ends gracefully.

**Reply by bk-cs** [reply]
It does not revoke before a request. Here is the exact logic that happens before _every_ request (if token is due to expire within 60 seconds, request a new token):
```powershell
if ($Script:Falcon.Expiration -le (Get-Date).AddSeconds(60)) { Request-FalconToken }
```
Although you can already reproduce it manually, you could make PSFalcon think the expiration is much lower by modifying the "expiration time" in `Public\oauth2.ps1`. Change this line from:
```powershell
$Script:Falcon.Expiration = (Get-Date).AddSeconds($Result.expires_in)
```
to:
```powershell
$Script:Falcon.Expiration = (Get-Date).AddSeconds(120)
```
That would cause your tokens to expire every two minutes, causing you to wait about a minute to see the failure happen.

I haven't received any reports about this sort of failure beyond yours, and (last I checked) there were about 1,500 different customers using PSFalcon daily. If there's something to change, I'd like to narrow it down and get it fixed, but if eliminating the proxy solves the issue, I have to think that the proxy is the reason.

**Reply by eddietorial** [reply]
Is it feasible to implement multiple retries (at least 3) on the request for a new token? Or maybe even overall for the module commands that may fail due to temporary and unavoidable network latency?

My observation over weeks of testing is that the default retry fails at the first attempt (in my environment).
A second attempt goes through without any errors.

As this only seems to be affecting my environment, I have put in new token requests within my script as a workaround.

However, another issue that I've frequently obeserved is that even if a target workstation is online, sometimes the `Invoke-FalconRtr` fails with a `"code":404,"message":"no successful hosts initialized on RTR"`.

So, I've put in a retry attempt similar to a solution described here
[https://stackoverflow.com/questions/45470999/powershell-try-catch-and-retry](https://stackoverflow.com/questions/45470999/powershell-try-catch-and-retry)

And, as expected, whenever I've encountered `"code":404,"message":"no successful hosts initialized on RTR"` again, it goes through successfully in the second attempt.

**Reply by bk-cs** [reply]
You can also use `-QueueOffline $true`  with `Start-FalconSession` to queue for offline hosts and avoid the `"code":404,"message":"no successful hosts initialized on RTR"` error (most of the time). This is a different problem than the token request, though.

I'm not sure it's appropriate to add an authorization token retry request to `Request-FalconToken`. This is the only report of token request issues that I've heard that isn't due to IP-based allowlisting or invalid credentials, and it seems restricted to your environment. The token request API also redirects if provided with the wrong hostname, so it might be a little tricky to add in a retry that still properly respects that redirection (i.e. don't retry if redirected, only if due to legitimate failure). I think the try/catch you're using in your script is the easiest way to handle it... unless you have other ideas?
